### DISCUSSION TITLE
Deno_Core: injecting a v8 variable into the runtime

### DISCUSSION THREAD

**Discussion by rwilcox** [denoland/deno#19232] [Q&A] [answered]
Howdy all!

I'm experimenting with deno_core to embed Deno/Javascript into my Rust program. I've noticed how you can add new functions to the runtime via the extensions mechanism, but I really want to add new _variables_.

Here's some code I think _might_ do this, but it has a Rust compile error about too many mutable borrows. ALSO: this is based half on snippets of code I found across the Internet, and half on conjecture. I _think_ it _might_ work... but the compiler errors are preventing me from knowing....

```rust

fn main() {

    let mut runtime = JsRuntime::new( Default::default() );
    let mut scope = runtime.handle_scope();

    let l1 = v8::String::new(&mut scope, "string value").unwrap();
    let l2 = v8::String::new(&mut scope, "our_value").unwrap();

    let variable_context = scope.get_current_context();
    let extra_bindings_obj = variable_context.get_extras_binding_object(&mut scope);
    extra_bindings_obj.set(&mut scope, l2.into(), l1.into());

    let res = runtime.execute_script_static("run filtering",
                                        "function do_echo(a) { return a } ; do_echo(JSON.stringify(Deno.core))").expect("err");

    let str = res.open(runtime.v8_isolate()).to_rust_string_lossy(&mut runtime.handle_scope());
    print!("from Javascript: {}", str);

}

```

This gives me the following compile error:
```
error[E0499]: cannot borrow `runtime` as mutable more than once at a time
  --> src/main.rs:16:15
   |
7  |       let mut scope = runtime.handle_scope();
   |                       ---------------------- first mutable borrow occurs here
...
16 |       let res = runtime.execute_script_static("run filtering",
   |  _______________^
17 | |                                         "function do_echo(a) { return a } ; do_echo(JSON.stringify(Deno.core))").expect("err");
   | |________________________________________________________________________________________________________________^ second mutable borrow occurs here
...
22 |   }
   |   - first borrow might be used here, when `scope` is dropped and runs the `Drop` code for type `HandleScope`

error[E0499]: cannot borrow `runtime` as mutable more than once at a time
  --> src/main.rs:19:24
   |
7  |     let mut scope = runtime.handle_scope();
   |                     ---------------------- first mutable borrow occurs here
...
19 |     let str = res.open(runtime.v8_isolate()).to_rust_string_lossy(&mut runtime.handle_scope());
   |                        ^^^^^^^^^^^^^^^^^^^^ second mutable borrow occurs here
...
22 | }
   | - first borrow might be used here, when `scope` is dropped and runs the `Drop` code for type `HandleScope`

error[E0499]: cannot borrow `runtime` as mutable more than once at a time
  --> src/main.rs:19:72
   |
7  |     let mut scope = runtime.handle_scope();
   |                     ---------------------- first mutable borrow occurs here
...
19 |     let str = res.open(runtime.v8_isolate()).to_rust_string_lossy(&mut runtime.handle_scope());
   |                                                                        ^^^^^^^^^^^^^^^^^^^^^^ second mutable borrow occurs here
...
22 | }
   | - first borrow might be used here, when `scope` is dropped and runs the `Drop` code for type `HandleScope`
```

I do see that methods like `runtime.handle_scope` do take a mutable self, and I do kind of understand that I am _actually_ modifying the Javascript heap, so I can see why those are mutable... but knowing this doesn't help my task.

TL; DR: My main goal is to inject some variables into my Rust JsRuntime. It's even fine if they are stand-alone globals (I found the extra_bindings method as a bit of a shortcut but I don't have to take it)

**Comment by bartlomieju** [comment]
Your code is almost fine - replace:
```
    let str = res.open(runtime.v8_isolate()).to_rust_string_lossy(&mut runtime.handle_scope());
```
with:
```
    let str = res.open(&mut scope).to_rust_string_lossy(&mut scope);
```
and it should work.

Since you already opened one handle scope there's really no point in this situation to create another. And for `res.open()` call you can also use existing `HandleScope` (because it internally defers to `Isolate`).

**Reply by rwilcox** [reply]
Ummmm _close_ actually! That gets rid of 2 of my errors. I'm only left with

```
error[E0499]: cannot borrow `runtime` as mutable more than once at a time
  --> src/main.rs:16:15
   |
7  |       let mut scope = runtime.handle_scope();
   |                       ---------------------- first mutable borrow occurs here
...
16 |       let res = runtime.execute_script_static("run filtering",
   |  _______________^
17 | |                                         "function do_echo(a) { return a } ; do_echo(JSON.stringify(Deno.core))").expect("err");
   | |________________________________________________________________________________________________________________^ second mutable borrow occurs here
18 |
19 |       let str = res.open(&mut scope).to_rust_string_lossy(&mut scope);
   |                          ---------- first borrow later used here
```

**Reply by bartlomieju** [reply]
In such situation it's probably gonna be:
```
fn main() {

    let mut runtime = JsRuntime::new( Default::default() );
    {
        let mut scope = runtime.handle_scope();

        let l1 = v8::String::new(&mut scope, "string value").unwrap();
        let l2 = v8::String::new(&mut scope, "our_value").unwrap();

        let variable_context = scope.get_current_context();
        let extra_bindings_obj = variable_context.get_extras_binding_object(&mut scope);
        extra_bindings_obj.set(&mut scope, l2.into(), l1.into());    
    }
    let res = runtime.execute_script_static("run filtering",
                                        "function do_echo(a) { return a } ; do_echo(JSON.stringify(Deno.core))").expect("err");

    let str = {
        let mut scope = runtime.handle_scope();
        res.open(&mut scope).to_rust_string_lossy(&mut scope)
    };
    print!("from Javascript: {}", str);

}
```

**Reply by rwilcox** [reply]
(I do very much appreciate the help so far) 

That compiles, but the I don't see `our_value` in the `Deno.core` object when it returns / when I print it out (which is where `get_extras_bindings_object` should put it, right?). I'm a bit worried that those v8::Strings will be deleted when they go out of scope?? But I'm a new Rustacean....

(I'm not tied to having it there, I just believe I need an object to set it on to...)

**Reply by bartlomieju** [reply, marked as answer]
To be honest I haven't used `get_extras_binding_object` so I can't say what it does. If you want to add additional properties to `Deno.core` you can just grab a reference to it:
```
    let global = variable_context.global(scope);
     let deno_str =
        v8::String::new(scope, "Deno").unwrap();
      let core_str =
        v8::String::new(scope, "core").unwrap();

      let deno_obj: v8::Local<v8::Object> = global
        .get(scope, deno_str.into())
        .unwrap()
        .try_into()
        .unwrap();
      let core_obj: v8::Local<v8::Object> = deno_obj
        .get(scope, core_str.into())
        .unwrap()
        .try_into()
        .unwrap();
```

> I'm a bit worried that those v8::Strings will be deleted when they go out of scope?? But I'm a new Rustacean....

What do you mean by deleted? If you use them to index objects that's fine - the same when you use them to assign something to an object. V8 will create a copy of these strings when necessary.

**Reply by rwilcox** [reply]
SO AWESOME! Thank you!!!!!! It works!
